Blazor RC2 Updates 8.0

guardrex opened the pull request:
Addresses #28161
Fixes #30372
Addresses #30002

# Notes

* Still a bit ***draft-y-ish*** at the moment. The whole PR will be edited on Monday morning.
* ~JS initializers coverage will be added on Monday.~ **For RTM**
* Release Notes updates will be made on Monday.
* The PR will merge and be taken live on Tuesday when RC2 releases.
* Adds a new section to the *Render Modes* article for Steve's exposition on ***custom shorthand render modes***. I thought that was 😎 ***Pretty Slick***&trade; 😎.
* **UPDATE for HttpContext**: ~I'll update and cross-link the passing tokens guidance on Monday.~ ***Done!*** 👍


The tracking issue has been updated ...

**Blazor .NET 8 tracking issue**
https://github.com/dotnet/AspNetCore.Docs/issues/28161

<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/advanced-scenarios.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/advanced-scenarios.md) | [ASP.NET Core Blazor advanced scenarios (render tree construction)](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/advanced-scenarios?branch=pr-en-us-30576) |
| [aspnetcore/blazor/call-web-api.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/call-web-api.md) | [Call a web API from an ASP.NET Core Blazor app](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/call-web-api?branch=pr-en-us-30576) |
| [aspnetcore/blazor/components/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/components/index.md) | [ASP.NET Core Razor components](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/index?branch=pr-en-us-30576) |
| [aspnetcore/blazor/components/integration.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/components/integration.md) | [aspnetcore/blazor/components/integration](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/integration?branch=pr-en-us-30576) |
| [aspnetcore/blazor/components/js-spa-frameworks.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/components/js-spa-frameworks.md) | [Use Razor components in JavaScript apps and SPA frameworks](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/js-spa-frameworks?branch=pr-en-us-30576) |
| [aspnetcore/blazor/components/prerender.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/components/prerender.md) | [Prerender ASP.NET Core Razor components](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/prerender?branch=pr-en-us-30576) |
| [aspnetcore/blazor/components/quickgrid.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/components/quickgrid.md) | [ASP.NET Core Blazor `QuickGrid` component](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/quickgrid?branch=pr-en-us-30576) |
| [aspnetcore/blazor/components/render-modes.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/components/render-modes.md) | [aspnetcore/blazor/components/render-modes](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?branch=pr-en-us-30576) |
| [aspnetcore/blazor/debug.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/debug.md) | [Debug ASP.NET Core apps](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/debug?branch=pr-en-us-30576) |
| [aspnetcore/blazor/file-downloads.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/file-downloads.md) | [ASP.NET Core Blazor file downloads](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/file-downloads?branch=pr-en-us-30576) |
| [aspnetcore/blazor/file-uploads.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/file-uploads.md) | [ASP.NET Core Blazor file uploads](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/file-uploads?branch=pr-en-us-30576) |
| [aspnetcore/blazor/forms-and-input-components.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/forms-and-input-components.md) | [ASP.NET Core Blazor forms and input components](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/forms-and-input-components?branch=pr-en-us-30576) |
| [aspnetcore/blazor/fundamentals/dependency-injection.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/fundamentals/dependency-injection.md) | [aspnetcore/blazor/fundamentals/dependency-injection](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?branch=pr-en-us-30576) |
| [aspnetcore/blazor/fundamentals/handle-errors.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/fundamentals/handle-errors.md) | [Handle errors in ASP.NET Core Blazor apps](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/handle-errors?branch=pr-en-us-30576) |
| [aspnetcore/blazor/fundamentals/routing.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/fundamentals/routing.md) | [ASP.NET Core Blazor routing and navigation](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?branch=pr-en-us-30576) |
| [aspnetcore/blazor/fundamentals/signalr.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/fundamentals/signalr.md) | [ASP.NET Core Blazor SignalR guidance](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/signalr?branch=pr-en-us-30576) |
| [aspnetcore/blazor/globalization-localization.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/globalization-localization.md) | [ASP.NET Core Blazor globalization and localization](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/globalization-localization?branch=pr-en-us-30576) |
| [aspnetcore/blazor/images.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/images.md) | [aspnetcore/blazor/images](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/images?branch=pr-en-us-30576) |
| [aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md) | [Call .NET methods from JavaScript functions in ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?branch=pr-en-us-30576) |
| [aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md) | [Call JavaScript functions from .NET methods in ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-javascript-from-dotnet?branch=pr-en-us-30576) |
| [aspnetcore/blazor/javascript-interoperability/import-export-interop.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/javascript-interoperability/import-export-interop.md) | [JavaScript `[JSImport]`/`[JSExport]` interop with ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?branch=pr-en-us-30576) |
| [aspnetcore/blazor/performance.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/performance.md) | [aspnetcore/blazor/performance](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance?branch=pr-en-us-30576) |
| [aspnetcore/blazor/project-structure.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/project-structure.md) | [ASP.NET Core Blazor project structure](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/project-structure?branch=pr-en-us-30576) |
| [aspnetcore/blazor/security/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/security/index.md) | [ASP.NET Core Blazor authentication and authorization](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/index?branch=pr-en-us-30576) |
| [aspnetcore/blazor/security/server/additional-scenarios.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/security/server/additional-scenarios.md) | [Server-side ASP.NET Core Blazor additional security scenarios](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/server/additional-scenarios?branch=pr-en-us-30576) |
| [aspnetcore/blazor/security/server/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/security/server/index.md) | [[Visual Studio](#tab/visual-studio)](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/server/index?branch=pr-en-us-30576) |
| [aspnetcore/blazor/state-management.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/state-management.md) | [aspnetcore/blazor/state-management](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/state-management?branch=pr-en-us-30576) |
| [aspnetcore/blazor/tooling.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/tooling.md) | [Tooling for ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/tooling?branch=pr-en-us-30576) |
| [aspnetcore/blazor/tutorials/build-a-blazor-app.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/blazor/tutorials/build-a-blazor-app.md) | [Build a Blazor todo list app](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/tutorials/build-a-blazor-app?branch=pr-en-us-30576) |
| [aspnetcore/mvc/views/razor.md](https://github.com/dotnet/AspNetCore.Docs/blob/6631467157af97bac5181b6e2756b48116c0ac7f/aspnetcore/mvc/views/razor.md) | [Razor syntax reference for ASP.NET Core](https://review.learn.microsoft.com/en-us/aspnet/core/mvc/views/razor?branch=pr-en-us-30576) |

</details>

> **Note**
> This table shows preview links for the 30 files with the most changes. For preview links for other files in this PR, select <strong>OpenPublishing.Build Details</strong> within [checks](https://github.com/dotnet/AspNetCore.Docs/pull/30576/checks).


<!-- PREVIEW-TABLE-END -->

danroth27 commented on aspnetcore/blazor/advanced-scenarios.md:
````diff
@@ -37,7 +37,7 @@ In <xref:Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder> methods wi
 ```razor
 @page "/built-content"
-@attribute [RenderModeServer]
+@rendermode RenderMode.InteractiveServer
````
This is fine for now, but I think we'll want to switch to the shorter form (`@rendermode InteractiveServer`) once we ship the updated template with the `using static` directive for `RenderMode`.

danroth27 commented on aspnetcore/blazor/components/event-handling.md:
```diff
@@ -142,6 +144,45 @@ For more information, see the following resources:
 
 * <xref:Microsoft.AspNetCore.Components.Web.EventHandlers> holds attributes to configure the mappings between event names and event argument types.
 
+:::moniker range=">= aspnetcore-8.0"
```
We don't typically call out support for individual events on elements like this. We could try to exhaustively document all of the events that Blazor has built-in support for, but that seems a bit excessive. Should we just move this to the release notes doc?

danroth27 commented on aspnetcore/blazor/components/index.md:
```diff
@@ -1478,9 +1476,6 @@ In the preceding code, the CSS selector, `#app`, indicates that the `App` compon
-<!-- UPDATE 8.0 Need to check on this: Is this still valid in a 
-     BW app, and is the render-mode still set in this way? -->
-
 MVC and Razor Pages apps can also use the [Component Tag Helper](xref:Microsoft.AspNetCore.Mvc.TagHelpers.ComponentTagHelper) to register statically-rendered Blazor WebAssembly root components:
```
Yeah, I'm not sure why this specifically refers to using the component tag helper with statically rendered Blazor WebAssembly components. The component tag helper supports the interactive server and WebAssembly render modes. It doesn't support Auto.

guardrex replied:
Let's discuss this further here. This whole section is specifically about rendering static root components. I thought it was correct in that context.

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -22,14 +22,14 @@ This article explains control of Razor component rendering in Blazor Web Apps, e
 Every component in a Blazor Web App adopts a *render mode* to determine the hosting model that it uses, where it's rendered, and whether or not it's interactive.
 
-The following table shows the available render modes for rendering Razor components in a Blazor Web App. To apply a render mode to a component use the `@rendermode` directive on the component instance or use the [`@attribute` Razor directive](xref:mvc/views/razor#attribute) with the corresponding render mode attribute in the component's Razor markup. Later in this article, examples are shown for  ...
+The following table shows the available render modes for rendering Razor components in a Blazor Web App. To apply a render mode to a component use the `@rendermode` directive on the component instance or in the component's Razor markup. Later in this article, examples are shown for each render mode scenario.
```
```suggestion
The following table shows the available render modes for rendering Razor components in a Blazor Web App. To apply a render mode to a component use the `@rendermode` directive on the component instance or on the component definition. Later in this article, examples are shown for each render mode scenario.
```

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -70,137 +72,122 @@ Example 1: The following `Program` file API adds services and configuration for
 ## Apply a render mode to a component instance
 
-To apply a render mode to a component instance use the [`@rendermode` Razor directive attribute](xref:mvc/views/razor#attribute) where the component is used.
+To apply a render mode to a component instance use the [`@rendermode` Razor directive](xref:mvc/views/razor#rendermode) where the component is used.
```
```suggestion
To apply a render mode to a component instance use the [`@rendermode` Razor directive attribute](xref:mvc/views/razor#rendermode) where the component is used.
```

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -70,137 +72,122 @@ Example 1: The following `Program` file API adds services and configuration for
 
 > [!NOTE]
-> During .NET 8 *Release Candidate 1*, use the following values:
+> The preceding syntax will be simplified at the release of .NET 8 in November to:
```
We could just tell people here that they can add the `using state` for `RenderMode` in their *_Imports.razor* to get the shorter syntax. That's what we plan to add in the Blazor Web App template.

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -278,17 +265,17 @@ In the following example, the render mode is set to WebAssembly with `@attribute
 If using the preceding component locally in a Blazor Web App, place the component in the client project's `Pages` folder. The client project is the solution's project with a name that ends in `.Client`. When the app is running, navigate to `/render-mode-3` in the browser's address bar.
 
-## Auto render mode
+## Automatic (Auto) render mode
```
Can we just stick with auto? I get that it probably violates some doc guideline of not abbreviating stuff, but it's what's shipping in the API.

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -490,3 +512,49 @@ The following example includes the assembly of the `DifferentAssemblyCounter` co
+
+<span aria-hidden="true">✔️</span><span class="visually-hidden">Valid:</span> `SomeMethod(new InteractiveServerRenderMode(prerender: false))`
+
+Therefore, disabling prerendering using the `@render-mode` directive requires the following example of an explicit Razor expression:
```
```suggestion
Therefore, disabling prerendering using the `@rendermode` directive requires the following example of an explicit Razor expression:
```

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -490,3 +512,49 @@ The following example includes the assembly of the `DifferentAssemblyCounter` co
+
+## Custom shorthand render modes
+
+Whatever follows an `@rendermode` directive must be a valid Razor expression that would make sense as a method parameter in C#.
```
I think this text can be simplified: The `@rendermode` directive takes a single parameter that is a static instance of type `IComponentRenderMode`. The `@rendermode` directive attribute can take any render mode instance (static or not). You can supply that render mode instance however you'd like. We provide the `RenderMode` static class with some predefined render modes for convenience, but you can of course create your own.

guardrex replied:
Yes, I like your (short and clear) lead in to the example. I think I can resolve this, and we'll take a look on the next commit.

danroth27 commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -490,3 +512,49 @@ The following example includes the assembly of the `DifferentAssemblyCounter` co
+
+Whatever follows an `@rendermode` directive must be a valid Razor expression that would make sense as a method parameter in C#.
+
+For example, consider how prerendering is disabled. The following expression isn't valid to pass to `@rendermode` because seeking to pass `InteractiveServer(prerender: false)` as a method parameter in `SomeMethod(InteractiveServer(prerender: false))` isn't a valid C# construct for a method parameter of `SomeMethod`.
```
This explanation isn't quite right: `InteractiveServer(prerender: false)` isn't valid because `InteractiveServer` isn't a method. It's a static property on the `RenderMode` static class.

guardrex replied:
For ...

> isn't valid because InteractiveServer isn't a method

... and what Steve provided, I'll work on this for the next commit.

For reference, Steve's remarks were ...

> Whatever follows @rendermode has to be a valid expression that would make sense as a method parameter in other code.
>
> For example you can’t do “@rendermode InteractiveServer(prerender: false)” because the following code isn’t valid:
>
> SomeMethod(InteractiveServer(prerender: false));
> 
> What would be valid is:
> 
> SomeMethod (new InteractiveServerRenderMode(prerender: false));
>
> and hence the syntax is “@rendermode new InteractiveServerRenderMode(prerender: false)”
>
> You’ll be able to write “@rendermode InteractiveServer” because we’ll have a “using static” for Microsoft.AspNetCore.Components.Web.RenderMode, and RenderMode.InteractiveServer exists as a static property, and hence:
>
> SomeMethod (InteractiveServer);
>
> … is equivalent to SomeMethod(RenderMode.InteractiveServer), and that’s all valid.

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
 
+:::moniker range=">= aspnetcore-8.0"
+
+## Enhanced navigation and form handling
```
This article is about forms, not navigation, so I think we can remove navigation from this heading.

```suggestion
## Enhanced form handling
```

I also think this section should go much later in the article, probably near the end. Enhanced form handling is totally optional and we don't want to scare off new users with these details right from the start. We can teach them all they need to know about form without any mention of enhanced form handling and then at the end let them know that they can ✨enhance✨ their Blazor forms by adding the appropriate attribute and explain what that does and how it interacts with redirects to non-Blazor pages.

danroth27 replied:
I think we can replace "enhanced navigation from a form" with "enhanced form handling" in this section.

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+
+## Enhanced navigation and form handling
+
+Examples adopt [enhanced navigation](xref:blazor/fundamentals/routing#enhanced-navigation-and-form-handling) for form POST requests with the `Enhance` parameter for `EditForm` forms or the `data-enhance` attribute for HTML forms (`<form>`):
```
Yeah, I think we should flip this around and have most of the examples not be enhanced and then just have one section that explains how you can choose to enhance your form if you want to.

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+>
+> This is the primary reason why enabling form POSTs with enhanced navigation requires an explicit gesture on each `EditForm` and `<form>` element. It isn't safe to assume that all forms in general should use enhanced posts because any given form might not target a Blazor endpoint.
+
+To disable enhanced navigation for a form:
```
```suggestion
To disable enhanced form handling:
```

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
````diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+</form>
+```
+
+<span aria-hidden="true">❌</span><span class="visually-hidden">Unsupported:</span> You can't set enhanced navigation on a form's ancestor element to enable enhanced navigation for the form.
````
```suggestion
<span aria-hidden="true">❌</span><span class="visually-hidden">Unsupported:</span> You can't set enhanced navigation on a form's ancestor element to enable enhanced form handling.
```

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+* For an `EditForm`, remove the `Enhance` parameter from the form element (or set it to `false`: `Enhance="false"`).
+* For an HTML `<form>`, remove the `data-enhance` attribute from form element (or set it to `false`: `data-enhance="false"`).
+
+Blazor's enhanced navigation and form handing may undo dynamic changes to the DOM if the updated content isn't part of the server rendering. To preserve the content of an element, use the `data-permanent` attribute.
```
I don't think we need to duplicate this content here. You've already got it covered in the Blazor routing doc, right? Can we just link to that for these details?

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
````diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+* For an `EditForm`, remove the `Enhance` parameter from the form element (or set it to `false`: `Enhance="false"`).
+* For an HTML `<form>`, remove the `data-enhance` attribute from form element (or set it to `false`: `data-enhance="false"`).
+
+Blazor's enhanced navigation and form handing may undo dynamic changes to the DOM if the updated content isn't part of the server rendering. To preserve the content of an element, use the `data-permanent` attribute.
+
+In the following example, the content of the `<div>` element is updated dynamically by a script when the page loads:
+
+```html
+<div data-permanent>
+    ...
+</div>
+```
+
+To disable enhanced navigation and form handling globally, see <xref:blazor/fundamentals/startup#enhanced-navigation-and-form-handling>.
+
+Once Blazor has started on the client, you can use the `enhancedload` event to listen for enhanced page updates. This allows for re-applying changes to the DOM that may have been undone by an enhanced page update.
+
+```javascript
+Blazor.addEventListener('enhancedload', () => console.log('Enhanced update!'));
+```
````
I don't think we need to duplicate this content here. You've already got it covered in the Blazor routing doc, right? Can we just link to that for these details?

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
````diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+If working with a standalone Blazor WebAssembly app, take ***either*** of the following approaches:
 
-* Change the render mode to `RenderModeWebAssembly`:
+* Change the render mode to `RenderMode.InteractiveWebAssembly`:
 
   ```diff
-  - @attribute [RenderModeServer]
-  + @attribute [RenderModeWebAssembly]
+  - @rendermode RenderMode.InteractiveServer
+  + @rendermode RenderMode.InteractiveWebAssembly
````
Standalone Blazor WebAssembly apps don't use the render modes. The whole app is running interactively on WebAssembly, just like it always has. All of the interactive forms in this doc should work just fine in a standalone Blazor WebAssembly app as long as the code doesn't make any assumptions about running on the server instead of the client.

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
 :::moniker range=">= aspnetcore-8.0"
 
-Components are configured for interactivity with server rendering. For a client-side experience in a Blazor Web App, change the render mode in the `@attribute` directive at the top of the component to either:
+Components are configured for interactivity with server rendering and enhanced navigation. For a client-side experience in a Blazor Web App, change the render mode in the `@rendermode` directive at the top of the component to either:
```
I think what we eventually want to do here is build up the user's understanding of form handling in Blazor by building off of concepts they probably already know, like the basics of forms in HTML. So, I think that means the initially sections should start with just a plan `<form>` (not `EditForm`) and then build up from there. First show how you can handle the form submission from the server using `@onsubmit`. Then introduce `EditForm` as a more convenient way to build a server based form. Then introduce enhanced form handling from the server. Then introduce interactive form handling. But we don't have to do this restructuring right away. It can wait until we get the .NET 8 stuff covered.

guardrex replied:
Very well, but it's probably best to reorganize the coverage after RC2 releases, perhaps late next week. I'll open an issue to make sure the task doesn't get lost when this PR merges.

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -3733,7 +3803,7 @@ If form processing fails because the component's form payload has exceeded the m
 <!-- UPDATE 8.0 Cross-link SSR -->
 
-Standard interactive HTML forms with server rendering are supported without using an <xref:Microsoft.AspNetCore.Components.Forms.EditForm>.
+Standard interactive HTML forms with server rendering are supported without using an <xref:Microsoft.AspNetCore.Components.Forms.EditForm>, including with ENHANCED NAV.
```
```suggestion
Standard interactive HTML forms with server rendering are supported without using an <xref:Microsoft.AspNetCore.Components.Forms.EditForm>, including with enhanced form handling.
```

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
```diff
@@ -3787,6 +3857,27 @@ Create a form using the normal HTML `<form>` tag and specify an `@onsubmit` hand
 
 To submit a form based on another element's DOM events, for example `oninput` or `onblur`, use JavaScript to submit the form ([`submit` (MDN documentation)](https://developer.mozilla.org/docs/Web/API/HTMLFormElement/submit)).
 
+Enhanced navigation is supported with the *explicit addition* of the `data-enhance` HTML attribute.
```
```suggestion
Enable enhanced form handling by adding the `data-enhance` HTML attribute to the form.
```

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
````diff
@@ -37,28 +37,90 @@ The <xref:Microsoft.AspNetCore.Components.Forms?displayProperty=fullName> namesp
+</div>
+```
+
+> [!WARNING]
+> ***Don't try to enhance form navigation for forms that POST to non-Blazor endpoints.***
+>
+> This is the primary reason why enabling form POSTs with enhanced navigation requires an explicit gesture on each `EditForm` and `<form>` element. It isn't safe to assume that all forms in general should use enhanced posts because any given form might not target a Blazor endpoint.
````
I think this warning might sound a bit too scary. I think all we really need to say is that enhanced forms only work with Blazor. Posting an enhanced form to non-Blazor endpoint will result in an error.

danroth27 commented on aspnetcore/blazor/forms-and-input-components.md:
````diff
@@ -3787,6 +3857,27 @@ Create a form using the normal HTML `<form>` tag and specify an `@onsubmit` hand
+</div>
+```
+
+> [!WARNING]
+> ***Don't try to enhance form navigation for forms that POST to non-Blazor endpoints.***
+>
+> This is the main reason why enabling form POSTs with enhanced navigation requires an explicit gesture on each `<form>` element. It isn't safe to assume that all forms in general should use enhanced posts because any given form might not target a Blazor endpoint.
````
Same comment about this warning as the one above. I want to avoid making it sound like enhanced form handling is somehow "unsafe".

danroth27 commented on aspnetcore/blazor/fundamentals/dependency-injection.md:
````diff
@@ -258,6 +258,24 @@ Prerequisites for constructor injection:
+public IMyService MyService { get; set; }
+```
+
+> [!NOTE]
+> The `@inject` Razor directive doesn't currently support keyed services, but this scenario is scheduled for the final release of .NET 8 in November.
````
```suggestion
> [!NOTE]
> The `@inject` Razor directive doesn't currently support keyed services.
```

danroth27 commented on aspnetcore/blazor/fundamentals/routing.md:
```diff
@@ -553,7 +553,76 @@ Pass `true` to the `forceLoad` parameter to ensure a full-page reload is always
+
+If the destination is a non-Blazor endpoint, enhanced navigation doesn't apply, and the client-side JavaScript retries as a full page load. This ensures no confusion to the framework about external pages that shouldn't be patched into an existing page.
+
+Form POST requests adopt [enhanced navigation](xref:blazor/fundamentals/routing#enhanced-navigation-and-form-handling) for form POST requests with the *explicit addition* of the `Enhance` parameter to `EditForm` forms or the *explicit addition* of the `data-enhance` attribute to the HTML form (`<form>`):
```
Is the "explicit addition" text trying to say that the `data-enhance` attribute can't be added dynamically, like with JS? If yes, should we just explicitly state that as an additional note?

```suggestion
To enable enhanced form handling, add the `Enhance` parameter to `EditForm` forms or the `data-enhance` attribute to HTML forms (`<form>`):
```

danroth27 commented on aspnetcore/blazor/fundamentals/routing.md:
````diff
@@ -553,7 +553,76 @@ Pass `true` to the `forceLoad` parameter to ensure a full-page reload is always
+</form>
+```
+
+Enhanced navigation isn't hierarchical and doesn't flow to child forms:
````
```suggestion
Enhanced form handling isn't hierarchical and doesn't flow to child forms:
```

danroth27 commented on aspnetcore/blazor/fundamentals/routing.md:
````diff
@@ -553,7 +553,76 @@ Pass `true` to the `forceLoad` parameter to ensure a full-page reload is always
+</div>
+```
+
+> [!WARNING]
+> ***Don't try to enhance form navigation for forms that POST to non-Blazor endpoints.***
+>
+> This is the main reason why enabling form POSTs with enhanced navigation requires an explicit gesture on each `EditForm` and `<form>` element and it doesn't flow down elements or to child forms. It isn't safe to assume that all forms in general should use enhanced posts because any given form might not target a Blazor endpoint.
````
See my comment on how to tone done this warning in the forms doc.

danroth27 commented on aspnetcore/blazor/security/includes/httpcontext.md:
```diff
@@ -1,5 +1,38 @@
+:::moniker range=">= aspnetcore-8.0"
+
+<xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> must be avoided when interactive rendering for the following reasons:
```
```suggestion
<xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> must be avoided with interactive rendering for the following reasons:
```

danroth27 commented on aspnetcore/blazor/security/includes/httpcontext.md:
```diff
@@ -1,5 +1,38 @@
+* The interface isn't implemented for interactive WebAssembly rendering.
+* For interactive server rendering, the implementation isn't guaranteed to exist.
+
+<xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> can be used for components that are statically rendered on the server. **However, we recommend avoid using it if possible.**
```
```suggestion
<xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> can be used for components that are statically rendered on the server. **However, we recommend to avoid using it if possible.**
```

danroth27 commented on aspnetcore/blazor/security/includes/httpcontext.md:
```diff
@@ -1,5 +1,38 @@
+:::moniker range=">= aspnetcore-8.0"
+
+<xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> must be avoided when interactive rendering for the following reasons:
+
+* The interface isn't implemented for interactive WebAssembly rendering.
+* For interactive server rendering, the implementation isn't guaranteed to exist.
```
```suggestion
<xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> must be avoided with interactive rendering because there isn't a valid `HttpContext` available.
```

danroth27 commented on aspnetcore/blazor/security/includes/httpcontext.md:
````diff
@@ -1,5 +1,38 @@
+public HttpContext? HttpContext { get; set; }
+```
+
+One caveat for using <xref:Microsoft.AspNetCore.Http.HttpContext> is that it won't compile in the `.Client` project of a Blazor Web App for WebAssembly rendering because the `.Client` project doesn't reference the [HTTP Abstractions package](https://www.nuget.org/packages/Microsoft.AspNetCore.Http.Abstractions). Even if the package reference is added to the `.Client` project, <xref:Microsoft.AspN ...
````
This seems like unnecessary much detail. We already stated above that you shouldn't try to access the `HttpContext` for interactive WebAssembly rendering.

```suggestion
```

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -25,12 +25,9 @@ To create a Blazor app on Windows, use the following guidance:
-     template survives in 8.0+ -->
-
 * Create a new project:
   * For a Blazor Web App experience (*recommended*), choose the **Blazor Web App** template. Select **Next**.
-  * For a Blazor WebAssembly experience, choose the **Blazor WebAssembly App** template, which includes demonstration code and Bootstrap, or the **Blazor WebAssembly App Empty** template without demonstration code and Bootstrap. Select **Next**.
+  * For a Blazor WebAssembly experience, choose the **Blazor WebAssembly Standalone App** template. Select **Next**.
```
The main point of these bullets seems to be to indicate the available Blazor templates. How about we just list the templates by name here and have a sentence about what each template is for?

```suggestion
* Create a new project using one of the available Blazor templates:
  * **Blazor Web App** (*recommended*): Creates a Blazor web app that supports both server and client-side rendering with rich interactivity. 
  * **Blazor WebAssembly Standalone App**: Creates a standalone client web app that can be deployed as a static site.
```

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -58,22 +55,32 @@ To create a Blazor app on Windows, use the following guidance:
+                **Interactive render mode** -->
 
-  If both the WebAssembly and Server render modes are selected, the template uses the Auto render mode. The Auto render mode initially uses the Server mode while the .NET app bundle and runtime are download to the browser. After the .NET WebAssembly runtime is activated, Auto switches to the WebAssembly render mode.
+  * **Interactivity type** dropdown list:
```
Just a heads up that the label for this dropdown will change to "Interactive render mode" for .NET 8 GA.

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -58,22 +55,32 @@ To create a Blazor app on Windows, use the following guidance:
 
-  If both the WebAssembly and Server render modes are selected, the template uses the Auto render mode. The Auto render mode initially uses the Server mode while the .NET app bundle and runtime are download to the browser. After the .NET WebAssembly runtime is activated, Auto switches to the WebAssembly render mode.
+  * **Interactivity type** dropdown list:
+    * Interactivity with server rendering is enabled by default with the **Server** option.
```
```suggestion
    * Interactive server rendering is enabled by default with the **Server** option.
```

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -58,22 +55,32 @@ To create a Blazor app on Windows, use the following guidance:
+  * **Interactivity type** dropdown list:
+    * Interactivity with server rendering is enabled by default with the **Server** option.
+    * To only enable interactivity with client rendering, select the **WebAssembly** option.
+    * To enable interactive server rendering followed by interactive WebAssembly rendering, select the **Auto (Server and WebAssembly)** (automatic) render mode option.
```
```suggestion
    * To enable both interactive server and client rendering and the ability to automatically switch between them at runtime, select the **Auto (Server and WebAssembly)** (automatic) render mode option.
```

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -58,22 +55,32 @@ To create a Blazor app on Windows, use the following guidance:
+    For more information, see <xref:blazor/components/render-modes>.
+
+  * **Interactivity location** dropdown list:
+    * **Per page/component**: The default sets up the app for interactivity set per page or per component.
```
```suggestion
    * **Per page/component**: The default sets up the app for interactivity per page or per component.
```

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -58,22 +55,32 @@ To create a Blazor app on Windows, use the following guidance:
+
+  * **Interactivity location** dropdown list:
+    * **Per page/component**: The default sets up the app for interactivity set per page or per component.
+    * **Global**: Selecting this option sets interactivity for entire app in the `App` component by specifying a render mode for the top-level `HeadOutlet` and `Routes` components. Setting the interactivity on these components propagates the interactivity to all of the child components in the app.
```
```suggestion
    * **Global**: Selecting this option sets up interactivity globally for entire app.
```

danroth27 commented on aspnetcore/blazor/tooling.md:
```diff
@@ -581,10 +613,10 @@ To disable SIMD, for example when targeting old browsers (on mobile devices), ad
 
 > [!IMPORTANT]
-> During 8.0 preview releases, disabling SIMD isn't supported. For more information, see [SIMD cannot be disabled (dotnet/runtime #89302)](https://github.com/dotnet/runtime/issues/89302).
+> Disabling SIMD isn't supported. For more information, see [SIMD cannot be disabled (dotnet/runtime #89302)](https://github.com/dotnet/runtime/issues/89302).
```
If disabling SIMD currently isn't supported should we remove this section? Why tell the user how to disable it and then immediate say it can't be disabled?

danroth27 commented on aspnetcore/blazor/tutorials/build-a-blazor-app.md:
```diff
@@ -113,10 +109,8 @@ The `-n|--name` option in the preceding command specifies the name of the new Ra
-
 * Add an `@page` Razor directive with a relative URL of `/todo`.
-* Add the `[RenderModeServer]` attribute. The attribute indicates that this component should be rendered on the server with client interactivity.
+* Set the interactive Server render mode for the component. The render mode indicates that this component should be rendered on the server with client interactivity.
```
```suggestion
* Set the interactive Server render mode for the component. The render mode indicates that this component should be rendered on the server with client interactivity.
```
```suggestion
* Enable interactivity on the page so that it isn't just statically rendered. The interactive server render mode enables the component to handle UI events from the server.
```

danroth27 commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -734,36 +734,59 @@ When set to `false` (default), whitespace in the rendered markup from Razor comp
+* `InteractiveAuto`: Initially applies interactive WebAssembly rendering using Blazor Server, and then applies interactive WebAssembly rendering using WebAssembly on subsequent visits after the Blazor bundle is downloaded.
 
-The following example sets the Server render mode to a `Dialog` component:
+For a component instance:
```
Technically `@rendermode` is both a Razor directive and a Razor directive attribute. The semantics are similar, but there are differences. The `@rendermode` directive is on the component definition, so the referenced render mode instance has to be static. The `@rendermode` directive attribute can take any render mode instance.

danroth27 reviewed (approved):
Looks good. I have some comments, but this is all in the right direction.

guardrex commented:
@danroth27 ... Reacted to feedback.

Sorry for so many initial commits to get your suggestions batched ... GH was throwing an error ***but wouldn't tell me which suggestion was failing***. I had to block them out in groups to narrow it down to the one that was choking the batch process.

I'm going to start the Release Notes update now but finish it on Monday morning.

I'll go back over this entire PR looking for further improvements and *gremlins* 😈 .

For some highlights .....................

### Support for dialog cancel and close events

... moving to Release Notes.

###  Component TH remarks

For ...

> MVC and Razor Pages apps can also use the \[Component Tag Helper](xref:Microsoft.AspNetCore.Mvc.TagHelpers.ComponentTagHelper) to register statically-rendered Blazor WebAssembly root components:

For >=8.0, I changed it to ...

> MVC and Razor Pages apps can also use the \[Component Tag Helper](xref:Microsoft.AspNetCore.Mvc.TagHelpers.ComponentTagHelper) to register interactive server and WebAssembly render mode components. The tag helper doesn't support Auto render mode components.

... but is the original remark still valid for <8.0?

### automatic (Auto)

... going back to just Auto.

### New section on custom shorthand render modes

... check again to see if the rewrite is correct. It's still might be rough, and you still might decide just to nix it. I liked it for a shorthand prerender approach, but it might not be worth it. As Steve said, how often will folks be disabling prerendering? Also, it's probably going to be more useful if more flags are added, but that might never happen.

### Forms article

Moved the *Enhanced navigation and form handling* section to the bottom and made the example forms not adopt it.

### `enhancedload` event to listen for enhanced page updates

... moved to the Routing topic and cross-linked from the Forms topic.

### Disabling SIMD

Holding coverage for .NET 9 ... I commented it out in place with a doc author comment.
